<a href="https://colab.research.google.com/github/jmatt724/InClassAssignment2_JLM/blob/main/Assignement3_JLM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [216]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Importing the dataset

In [217]:
dataset = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/newhousing.csv")

## Preprocessing

In [218]:
dataset.columns

Index(['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'mainroad',
       'guestroom', 'basement', 'hotwaterheating', 'airconditioning',
       'parking', 'prefarea', 'semi-furnished', 'unfurnished',
       'areaperbedroom', 'bbratio'],
      dtype='object')

In [219]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   price            545 non-null    int64  
 1   area             545 non-null    int64  
 2   bedrooms         545 non-null    int64  
 3   bathrooms        545 non-null    int64  
 4   stories          545 non-null    int64  
 5   mainroad         545 non-null    int64  
 6   guestroom        545 non-null    int64  
 7   basement         545 non-null    int64  
 8   hotwaterheating  545 non-null    int64  
 9   airconditioning  545 non-null    int64  
 10  parking          545 non-null    int64  
 11  prefarea         545 non-null    int64  
 12  semi-furnished   545 non-null    int64  
 13  unfurnished      545 non-null    int64  
 14  areaperbedroom   545 non-null    float64
 15  bbratio          545 non-null    float64
dtypes: float64(2), int64(14)
memory usage: 68.2 KB


### No imputation needed

## No Dummy variables needed

In [220]:
X = dataset.drop('price', axis = 1)
y = dataset['price']

### Feature Scaling

In [221]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_scaled = sc.fit_transform(X)

### Linear Regression

In [222]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42, shuffle = True)

In [223]:
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression()

In [224]:
lin_reg.score(X_test, y_test)

0.6814896416143947

In [225]:
lin_reg.score(X_train, y_train)

0.6760953066606615

## Random Forrest Regression

In [226]:
from sklearn.ensemble import RandomForestRegressor

rnd_reg = RandomForestRegressor()

rnd_reg.fit(X_train, y_train)

RandomForestRegressor()

In [227]:
rnd_reg.score(X_test, y_test)

0.7125164649830784

In [228]:
rnd_reg.score(X_train, y_train)

0.9363466175333

### Random Forest is overfitting

## PCA with Linear Regression

In [229]:
from sklearn.preprocessing import StandardScaler

sc_x = StandardScaler()
sc_y = StandardScaler()
X_scaled = sc_x.fit_transform(X)
y_scaled = sc_y.fit_transform(dataset[['price']])

In [230]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 2)
pca.fit(X_scaled)
x_pca = pca.transform(X_scaled)

In [231]:
X_train, X_test, y_train, y_test = train_test_split(x_pca, y_scaled, test_size = 0.20, random_state=42, shuffle = True)

In [232]:
pca_lin_reg = LinearRegression()
pca_lin_reg.fit(X_train, y_train)

LinearRegression()

In [233]:
pca_lin_reg.score(X_test, y_test)

0.6444402044504696

In [236]:
pca_lin_reg.score(X_train, y_train)

0.6480907854037373

## PCA with Random Forest

In [234]:
pca_rnd_reg = RandomForestRegressor()
pca_rnd_reg.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


RandomForestRegressor()

In [235]:
pca_rnd_reg.score(X_test, y_test)

0.5985646955959366

In [238]:
pca_rnd_reg.score(X_train, y_train)

0.9372982415088245

# For my experiment I found that Random Forest had the best score with 71% but it overfits. Therefore the best model that doesn't overfit is Linear Regression with 68%